### PR TITLE
Metadata wrong exportation when containing invalid path character

### LINKED ISSUE #14758: Maps with slashes in metadata are exported incorrectly
**Describe the bug:** When a map contains slash character anywhere in artist/title/difficulty metadata, after exporting it, it's unimportable in stable. It looks like slashes aren't sanitized so .osu files end up being in subfolders.

Example maps that fail to import after exporting in lazer:
https://osu.ppy.sh/beatmapsets/423194#osu/914570
https://osu.ppy.sh/beatmapsets/855677#osu/1787848

**Screenshots or videos showing encountered issue:**
image
image

**osu!lazer version:** 2021.907.0-lazer

**Logs:** not needed

### PR DESCRIPTION
**Pull request by Youssef-Hammad**
"Maps with slashes in metadata are exported incorrectly"
Replaced all incompatible characters in maps' metadata with empty string right before exporting, testing with importing the maps to and from osu!lazer and stable (kept going back and forth with the import/export)

Closes #14758

### CODE REVIEW

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by LittleEndu** [comment]
Pull request titles are used in the changelog, so it needs to be changed into what issue exactly it is fixing. Also you should mention "Closes #14758" in the pull request description so the issue will get automatically closed once the PR is merged.

File: osu.Game/Beatmaps/BeatmapManager.cs
Comment (Youssef-Hammad) [reply osu.Game/Beatmaps/BeatmapManager.cs]: did you mean that I use the one in ArchiveModelManager.cs ? Because it's a private function, I went ahead and wrote those 2 lines

File: osu.Game/Beatmaps/BeatmapManager.cs
Comment (Susko3) [inline osu.Game/Beatmaps/BeatmapManager.cs:268]: This shouldn't be in the diff, please format your files with your code editor before committing/making a pull request.

File: osu.Game/Beatmaps/BeatmapManager.cs
Comment (Susko3) [inline osu.Game/Beatmaps/BeatmapManager.cs:272]: I'd write something like:

```suggestion
                    // remove incompatible characters from path
```

If you look at the other comments, they all use the present tense.

File: osu.Game/Beatmaps/BeatmapManager.cs
Comment (Susko3) [inline osu.Game/Beatmaps/BeatmapManager.cs:274]: Again, this needs to be properly formatted.
```suggestion
                    foreach (char c in Path.GetInvalidFileNameChars())
                        beatmapInfo.Path = beatmapInfo.Path.Replace(c.ToString(), "");
```

**Review by Susko3 — CHANGES_REQUESTED** [review]

File: osu.Game/Beatmaps/BeatmapManager.cs
Comment (Youssef-Hammad) [reply osu.Game/Beatmaps/BeatmapManager.cs]: All changes done, check now

File: osu.Game/Beatmaps/BeatmapManager.cs
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapManager.cs:273]: This already exists in `ArchiveModelManager`. Please use that function instead.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Beatmaps/BeatmapManager.cs
Comment (Youssef-Hammad) [reply osu.Game/Beatmaps/BeatmapManager.cs:273]: If you're referring to this one, it's a private function that's why I couldn't use it. What do you suggest?
https://github.com/ppy/osu/blob/f5a5b87cf6e633e237b7595b1ef9cbf3dffe3da7/osu.Game/Database/ArchiveModelManager.cs#L896-L901

Comment (peppy) [reply osu.Game/Beatmaps/BeatmapManager.cs:273]: make it not private. in this case `protected static` would likely work fine.

Comment (Youssef-Hammad) [reply osu.Game/Beatmaps/BeatmapManager.cs]: Done

Comment (peppy) [reply osu.Game/Beatmaps/BeatmapManager.cs]: Please don't self-resolve reviews. We will resolve them once we have re-checked the issue. Also there's no need to reply with "done" - github alerts reviewers with notifications when a change is pushed.

File: osu.Game/Database/ArchiveModelManager.cs
Comment (peppy) [inline osu.Game/Database/ArchiveModelManager.cs:896]: naming is wrong. please check the README for contribution guidelines, which include the checks you should be running before pushing.

Comment (Youssef-Hammad) [reply osu.Game/Database/ArchiveModelManager.cs:896]: How do I run checks before pushing? I thought the pending checks are awaited approval to run by contributors.
Sorry for the silly questions, this is my first time contributing to an open-source project so I'm getting exposed to a part of github I haven't usually dealt with on my day-to-day use

Comment (peppy) [reply osu.Game/Database/ArchiveModelManager.cs:896]: You can check the documentation here: https://github.com/ppy/osu#code-analysis

**Comment by peppy** [comment]
Not sure on this solution.

- Relying on `Save` to be called before `Export`, or the same issue will be encountered. I think we also want this to be applied to some degree on beatmap *import*, before the incorrect values are written to the database?
- Missing tests

**Comment by Youssef-Hammad** [comment]
> Not sure on this solution.
> 
> * Relying on `Save` to be called before `Export`, or the same issue will be encountered. I think we also want this to be applied to some degree on beatmap _import_, before the incorrect values are written to the database?
> * Missing tests

* I implemented it in `Save` since the file path is constructed there, I've tried exporting the map "Systematic Love" mentioned in the issue using both stable and lazer and ended up with the same result after this solution. The import I think would be a separate issue? I don't see how the import will bug unless the .osz was manually tampered with

* I'm still trying to add the tests, will include them in the next push (hopefully)

**Comment by PercyDan54** [comment]
> > Not sure on this solution.
> > 
> > * Relying on `Save` to be called before `Export`, or the same issue will be encountered. I think we also want this to be applied to some degree on beatmap _import_, before the incorrect values are written to the database?
> > * Missing tests
> 
> * I implemented it in `Save` since the file path is constructed there, I've tried exporting the map "Systematic Love" mentioned in the issue using both stable and lazer and ended up with the same result after this solution. The import I think would be a separate issue? I don't see how the import will bug unless the .osz was manually tampered with
> * I'm still trying to add the tests, will include them in the next push (hopefully)
> 
> Edit: How exactly can I create an automated test for this case? The process is very manual (Download beatmap -> Import -> Export -> Look at the file structure to see if there's a directory other than sb or is overall matching to exports coming out of stable)
> I switched to Rider to run InspectCode and everything seems fine

https://github.com/ppy/osu/blob/5c7fe5dde0fdc861f48d1741c01d8312e52ce3b3/osu.Game.Tests/Beatmaps/IO/ImportBeatmapTest.cs#L139

**Comment by Youssef-Hammad** [comment]
I'm not entirely sure if the test in my last commit is valid, it's a first for me writing a non-trivial automated test. Although it does seem logical and the test pass

File: osu.Game.Tests/Beatmaps/IO/ImportBeatmapTest.cs
Comment (peppy) [inline osu.Game.Tests/Beatmaps/IO/ImportBeatmapTest.cs:139]: I have no idea what this test method name means

File: osu.Game.Tests/Beatmaps/IO/ImportBeatmapTest.cs
Comment (peppy) [inline osu.Game.Tests/Beatmaps/IO/ImportBeatmapTest.cs:151]: Also I can't understand what is being tested here after three reads. Start by adding inline comments explaining what your intentions are

**Comment by Youssef-Hammad** [comment]
This looks like a more logical testing approach. I compared the content of the original .osz with the .osz after the beatmapInfo.Path for all beatmaps undergoes the same construction (and filtering of invalid characters) as done in BeatmapManager.cs Save() and see if the content is identical.

File: osu.Game.Tests/Beatmaps/IO/ImportBeatmapTest.cs
Comment (peppy) [inline osu.Game.Tests/Beatmaps/IO/ImportBeatmapTest.cs:151]: this will be my last comment on this review but this is wrong. you can't just assume a temp path directory location like this. follow other examples, read existing code, write good code.

File: osu.Game.Tests/Beatmaps/IO/ImportBeatmapTest.cs
Comment (bdach) [inline osu.Game.Tests/Beatmaps/IO/ImportBeatmapTest.cs:189]: this is unacceptable. this is explicitly bypassing *all of the code this is supposed to be testing* by reimplementing both the problem scenario and the fix in the test. therefore this test, in fact, tests *absolutely nothing of use*. i'm 99% certain i could revert your _actual_ fix and still have this test pass.

i'm sorry but i'm not willing to review this pull any more than i already have, either. a change as simple as this should never have undergone so many rounds of grueling review.

**Review by bdach — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
